app/models/oidc_account.rb

```diff
@@ -8,9 +8,13 @@ class OidcAccount < ApplicationRecord
   validates :uid, presence: true, uniqueness: true
 
   def self.link(user, auth)
-    upsert_all(
-      [{user_id: user.id, provider: auth.provider, uid: auth.uid}],
-      unique_by: [:user_id]
-    )
+    attributes = {
+      user_id: user.id,
+      provider: auth.provider,
+      uid: auth.uid,
+      token: auth.dig(:credentials, :token),
+      refresh_token: auth.dig(:credentials, :refresh_token),
```

Comment:
It appears that the absence of these tokens indicates it's expired. 
I have to admit ignorance as to how the auth session works. Are these tokens in a browser cookie that expires? or is Devise actively checking the currency of these tokens on each request? And is it checking against a timestamp, or an external source (and therefore making an extra round trip for every request)?

Reply:
I forgot to answer this earlier.

The tokens are now stored in our database and we can use them when calling other DFC APIs. The token contains an expiry timestamp and the token is signed so that it can't be changed. But the receiving API may also check with the auth server if the token is still valid. It could have been revoked. OFN doesn't do that though. Tokens are only valid for half an hour and we trust that timestamp. Attack scenario: the lescommuns server could get compromised and the attacker could generate many access tokens with future dates to use at any time. If this is noticed and the tokens are revoked, OFN wouldn't realise. But in such a severe case of compromise, we should probably re-generate the keys of the lescommuns server and start from scratch, invalidating all old tokens that way.

But back to what OFN is doing. So now we got an access token and a refresh token. The access token will be used to import a DFC product catalog, for example. If that fails then we can use the refresh token to get a new access token from lescommuns. If that fails as well then the refresh token is invalid we should delete it from our database. None of this is implemented yet and that's why you don't know it yet. :wink: 

If the user is triggering the product import then we can redirect them straight to lescommuns to log in again but if it happens offline, like placing a DFC order then we don't have a user to redirect anywhere. Deleting the refresh token will be a way to remember that it's invalid. We probably want to send an email to the user to let them know. They can then come to the OIDC settings page and login again so that we get new tokens to do the job.

This makes me think that we may need a way to restart failed jobs in this case. We'll see.